File: docs/legacy(old PACE)/C#/PaceAPISample/sample/FindObjectsAggregate.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using efipaceservices;

namespace Pace_Web_Service_SDK.sample
{
    class FindObjectsAggregate
    {
        public static void Run()
        {
            FindObjectsPortType findObjectsHttpBinding = PaceClient.getFindObjectsHttpBinding();

            //create the Job ValueObjectDescriptor for first batch of 10 records
            ValueObjectDescriptor jobDescriptor = createJobDescriptor(0, 10);

            // Open Job aggregate value object loader Sample
            ValueObjectsGroup jobVOs = findObjectsHttpBinding.loadValueObjects(jobDescriptor);
            printValueObjectGroupDetail(jobVOs, 0);

            //if there are more records, print the next batch.
            //Note:- you may use this pagination feature on any level of ValueObjectDescriptors
            if (jobVOs.totalRecords > 10)
            {
                printValueObjectGroupDetail(findObjectsHttpBinding.loadValueObjects(createJobDescriptor(11, 10)), 0);
            }
            
        }

        private static void printValueObjectGroupDetail(ValueObjectsGroup voGroup, int level)
        {
            String padding = getPadding(level);

            Console.WriteLine(padding + "Total " + voGroup.objectName + " available " + voGroup.totalRecords);

            ValueObject[] vos = voGroup.valueObjects;

            foreach (ValueObject valueObject in vos)
            {
                ValueField[] fields = valueObject.fields;
                Console.WriteLine(padding + "*** " + valueObject.objectName + " PK:" + valueObject.primaryKey + " ***");

                foreach (ValueField valueField in fields)
                {
                    Console.WriteLine(padding + valueObject.objectName + " Field: Name=" + valueField.name + ", Type=" + valueField.type + ",Value=" + valueField.value + ", XPath= " + valueField.xpath);
                }

                ValueObjectsGroup[] children = valueObject.children;

                foreach (ValueObjectsGroup child in children)
                {
                    printValueObjectGroupDetail(child, level + 1);
                }
                Console.WriteLine(padding + "*** " + valueObject.objectName + " PK:" + valueObject.primaryKey + " ***");
            }
        }

        private static String getPadding(int level)
        {
            String paddingString = "";
            for (int i = 0; i < level; i++)
            {
                paddingString += "\t";
            }
            return paddingString.ToString();
        }


        private static ValueObjectDescriptor createJobDescriptor(int offset, int limit)
        {
            // create the root ValueObjectDescriptor
            ValueObjectDescriptor jobDescriptor = new ValueObjectDescriptor();
            // set descriptor properties for Job lookup
            jobDescriptor.objectName = "Job";
            jobDescriptor.offset = offset;
            jobDescriptor.limit = limit;
            jobDescriptor.xpathFilter = "adminStatus/@openJob";

            // create fields to lookup
            List<FieldDescriptor> jobFieldDescriptors = new List<FieldDescriptor>();

            FieldDescriptor fieldDescriptor2 = new FieldDescriptor();
            fieldDescriptor2.name = "description";
            fieldDescriptor2.xpath = "@description";
            jobFieldDescriptors.Add(fieldDescriptor2);


            FieldDescriptor fieldDescriptor3 = new FieldDescriptor();
            fieldDescriptor3.name = "promiseDate";
            fieldDescriptor3.xpath = "@promiseDate";
            jobFieldDescriptors.Add(fieldDescriptor3);


            FieldDescriptor fieldDescriptor4 = new FieldDescriptor();
            fieldDescriptor4.name = "promiseTime";
            fieldDescriptor4.xpath = "@promiseTime";
            jobFieldDescriptors.Add(fieldDescriptor4);

            // add fields to job descriptor
            jobDescriptor.fields = jobFieldDescriptors.ToArray();

            // define sorts to be used
            XPathDataSort[] sort = new XPathDataSort[2];

            XPathDataSort xPathDataSort1 = new XPathDataSort();
            xPathDataSort1.descending = false;
            xPathDataSort1.xpath = "customer/@custName";

            XPathDataSort xPathDataSort2 = new XPathDataSort();
            xPathDataSort2.descending = true;
            xPathDataSort2.xpath = "@description";

            sort[0] = xPathDataSort1;
            sort[1] = xPathDataSort2;

            jobDescriptor.xpathSorts = sort;

            // create the JobPart descriptor that we want to pull with the Jobs
            ValueObjectDescriptor jobPartDescriptor = new ValueObjectDescriptor();
            jobPartDescriptor.objectName = "JobPart";
            jobPartDescriptor.offset = 0;
            jobPartDescriptor.limit = 10;
            // xpath is relative to this object. Leave null if you do not need a filter at this level
            //The parent filter will automatically be applied for the children
            //For eg:- the below filter will always be evaluated in context of the selected Job
            jobPartDescriptor.xpathFilter = "productionStatus/@openJob";

            // create fields to lookup
            List<FieldDescriptor> jobPartFieldDescriptors = new List<FieldDescriptor>();

            FieldDescriptor jobPartfieldDescriptor1 = new FieldDescriptor();
            jobPartfieldDescriptor1.name = "job";
            jobPartfieldDescriptor1.xpath = "@job";
            jobPartFieldDescriptors.Add(jobPartfieldDescriptor1);

            FieldDescriptor jobPartfieldDescriptor2 = new FieldDescriptor();
            jobPartfieldDescriptor2.name = "jobPart";
            jobPartfieldDescriptor2.xpath = "@jobPart";
            jobPartFieldDescriptors.Add(jobPartfieldDescriptor2);


            FieldDescriptor jobPartfieldDescriptor3 = new FieldDescriptor();
            jobPartfieldDescriptor3.name = "productionStatusDesc";
            jobPartfieldDescriptor3.xpath = "productionStatus/@description";
            jobPartFieldDescriptors.Add(jobPartfieldDescriptor3);

            FieldDescriptor jobPartfieldDescriptor4 = new FieldDescriptor();
            jobPartfieldDescriptor4.name = "qtyToMfg";
            jobPartfieldDescriptor4.xpath = "@qtyToMfg";
            jobPartFieldDescriptors.Add(jobPartfieldDescriptor4);

            jobPartDescriptor.fields = jobPartFieldDescriptors.ToArray();


            // create the JobShipment descriptor that we want to pull with the Jobs
            ValueObjectDescriptor jobShipmentDescriptor = new ValueObjectDescriptor();
            jobShipmentDescriptor.objectName = "JobShipment";
            jobShipmentDescriptor.offset = 0;
            jobShipmentDescriptor.limit = 10;

            // create fields to lookup
            List<FieldDescriptor> jobShipmentFieldDescriptors = new List<FieldDescriptor>();

            FieldDescriptor jobShipmentFieldDescriptor1 = new FieldDescriptor();
            jobShipmentFieldDescriptor1.name = "id";
            jobShipmentFieldDescriptor1.xpath = "@id";

            FieldDescriptor jobShipmentFieldDescriptor2 = new FieldDescriptor();
            jobShipmentFieldDescriptor2.name = "job";
            jobShipmentFieldDescriptor2.xpath = "@job";

            FieldDescriptor jobShipmentFieldDescriptor3 = new FieldDescriptor();
            jobShipmentFieldDescriptor3.name = "jobPart";
            jobShipmentFieldDescriptor3.xpath = "@jobPart";

            FieldDescriptor jobShipmentFieldDescriptor4 = new FieldDescriptor();
            jobShipmentFieldDescriptor4.name = "Shipment Type";
            jobShipmentFieldDescriptor4.xpath = "@shipmentType/@description";

            // add fields to jobpart descriptor
            jobShipmentDescriptor.fields = jobShipmentFieldDescriptors.ToArray();

            // set the child in the jobdescriptor
            jobDescriptor.children = new ValueObjectDescriptor[] { jobPartDescriptor, jobShipmentDescriptor };

            return jobDescriptor;
        }

    }
}

File: docs/legacy(old PACE)/unsupported/VB.NET/PaceWebServiceSDK/sample/FindObjectsAggregate.vb
Imports PaceWebServiceSDK.efipaceservices

Public Class FindObjectsAggregate

    Public Shared Sub Run()
        Dim findObjectsHttpBinding As FindObjectsHttpBinding = PaceClient.getFindObjectsHttpBinding()

        'create the Job ValueObjectDescriptor for first batch of 10 records
        Dim jobDescriptor As ValueObjectDescriptor = createJobDescriptor(0, 10)

        ' Open Job aggregate value object loader Sample
        Dim jobVOs As ValueObjectsGroup = findObjectsHttpBinding.loadValueObjects(jobDescriptor)
        printValueObjectGroupDetail(jobVOs, 0)

        'if there are more records, print the next batch.
        'Note:- you may use this pagination feature on any level of ValueObjectDescriptors
        If jobVOs.totalRecords > 10 Then
            printValueObjectGroupDetail(findObjectsHttpBinding.loadValueObjects(createJobDescriptor(11, 10)), 0)
        End If
    End Sub

    Private Shared Sub printValueObjectGroupDetail(ByVal voGroup As ValueObjectsGroup, ByVal level As Integer)
        Console.WriteLine("Total " + voGroup.objectName + " available " + voGroup.totalRecords.ToString)

        Dim vos As ValueObject() = voGroup.valueObjects

        For Each valueObject As ValueObject In vos
            Dim fields As ValueField() = valueObject.fields
            Console.WriteLine("*** " + valueObject.objectName + " PK:" + valueObject.primaryKey + " ***")

            For Each valueField As ValueField In fields
                Console.WriteLine(valueObject.objectName + " Field: Name=" + valueField.name + ", Type=" + valueField.type + ",Value=" + valueField.value + ", XPath= " + valueField.xpath)
            Next

            Dim children As ValueObjectsGroup() = valueObject.children

            For Each child As ValueObjectsGroup In children
                printValueObjectGroupDetail(child, level + 1)
            Next
            Console.WriteLine("*** " + valueObject.objectName + " PK:" + valueObject.primaryKey + " ***")
        Next
    End Sub


    Private Shared Function createJobDescriptor(ByVal offset As Integer, ByVal limit As Integer) As ValueObjectDescriptor

        ' create the root ValueObjectDescriptor
        Dim jobDescriptor As ValueObjectDescriptor = New ValueObjectDescriptor()
        ' set descriptor properties for Job lookup
        jobDescriptor.objectName = "Job"
        jobDescriptor.offset = offset
        jobDescriptor.limit = limit
        jobDescriptor.xpathFilter = "@dateSetup = date( 2012, 08, 01 )"

        ' create fields to lookup
        Dim jobFieldDescriptors As New List(Of FieldDescriptor)()

        Dim fieldDescriptor2 As FieldDescriptor = New FieldDescriptor()
        fieldDescriptor2.name = "description"
        fieldDescriptor2.xpath = "@description"
        jobFieldDescriptors.Add(fieldDescriptor2)


        Dim fieldDescriptor3 As FieldDescriptor = New FieldDescriptor()
        fieldDescriptor3.name = "promiseDate"
        fieldDescriptor3.xpath = "@promiseDate"
        jobFieldDescriptors.Add(fieldDescriptor3)


        ' add fields to job descriptor
        jobDescriptor.fields = jobFieldDescriptors.ToArray()

        ' define sorts to be used
        Dim sort As XPathDataSort() = New XPathDataSort(1) {}

        Dim xPathDataSort1 As XPathDataSort = New XPathDataSort()
        xPathDataSort1.descending = False
        xPathDataSort1.xpath = "customer/@custName"

        Dim xPathDataSort2 As XPathDataSort = New XPathDataSort()
        xPathDataSort2.descending = True
        xPathDataSort2.xpath = "@description"

        sort(0) = xPathDataSort1
        sort(1) = xPathDataSort2

        jobDescriptor.xpathSorts = sort

        ' create the JobPart descriptor that we want to pull with the Jobs
        Dim jobPartDescriptor As ValueObjectDescriptor = New ValueObjectDescriptor()
        jobPartDescriptor.objectName = "JobPart"
        jobPartDescriptor.offset = 0
        jobPartDescriptor.limit = 10
        ' xpath is relative to this object. Leave null if you do not need a filter at this level
        'The parent filter will automatically be applied for the children
        'For eg:- the below filter will always be evaluated in context of the selected Job
        jobPartDescriptor.xpathFilter = "productionStatus/@openJob"

        ' create fields to lookup
        Dim jobPartFieldDescriptors As New List(Of FieldDescriptor)()

        Dim jobPartfieldDescriptor1 As FieldDescriptor = New FieldDescriptor()
        jobPartfieldDescriptor1.name = "job"
        jobPartfieldDescriptor1.xpath = "@job"
        jobPartFieldDescriptors.Add(jobPartfieldDescriptor1)

        Dim jobPartfieldDescriptor2 As FieldDescriptor = New FieldDescriptor()
        jobPartfieldDescriptor2.name = "jobPart"
        jobPartfieldDescriptor2.xpath = "@jobPart"
        jobPartFieldDescriptors.Add(jobPartfieldDescriptor2)


        Dim jobPartfieldDescriptor3 As FieldDescriptor = New FieldDescriptor()
        jobPartfieldDescriptor3.name = "productionStatusDesc"
        jobPartfieldDescriptor3.xpath = "productionStatus/@description"
        jobPartFieldDescriptors.Add(jobPartfieldDescriptor3)

        Dim jobPartfieldDescriptor4 As FieldDescriptor = New FieldDescriptor()
        jobPartfieldDescriptor4.name = "qtyToMfg"
        jobPartfieldDescriptor4.xpath = "@qtyToMfg"
        jobPartFieldDescriptors.Add(jobPartfieldDescriptor4)

        jobPartDescriptor.fields = jobPartFieldDescriptors.ToArray()


        ' create the JobShipment descriptor that we want to pull with the Jobs
        Dim jobShipmentDescriptor As ValueObjectDescriptor = New ValueObjectDescriptor()
        jobShipmentDescriptor.objectName = "JobShipment"
        jobShipmentDescriptor.offset = 0
        jobShipmentDescriptor.limit = 10


        ' create fields to lookup
        Dim jobShipmentFieldDescriptors As New List(Of FieldDescriptor)()

        Dim jobShipmentFieldDescriptor1 As FieldDescriptor = New FieldDescriptor()
        jobShipmentFieldDescriptor1.name = "id"
        jobShipmentFieldDescriptor1.xpath = "@id"

        Dim jobShipmentFieldDescriptor2 As FieldDescriptor = New FieldDescriptor()
        jobShipmentFieldDescriptor2.name = "job"
        jobShipmentFieldDescriptor2.xpath = "@job"

        Dim jobShipmentFieldDescriptor3 As FieldDescriptor = New FieldDescriptor()
        jobShipmentFieldDescriptor3.name = "jobPart"
        jobShipmentFieldDescriptor3.xpath = "@jobPart"

        Dim jobShipmentFieldDescriptor4 As FieldDescriptor = New FieldDescriptor()
        jobShipmentFieldDescriptor4.name = "Shipment Type"
        jobShipmentFieldDescriptor4.xpath = "@shipmentType/@description"

        ' add fields to jobpart descriptor
        jobShipmentDescriptor.fields = jobShipmentFieldDescriptors.ToArray()

        ' set the child in the jobdescriptor
        jobDescriptor.children = New ValueObjectDescriptor() {jobPartDescriptor, jobShipmentDescriptor}

        Return jobDescriptor
    End Function
End Class
